cli/azd/pkg/ext/hooks.go

```diff
@@ -0,0 +1,379 @@
+package ext
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/exec"
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
+	"github.com/azure/azure-dev/cli/azd/pkg/output"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/bash"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/powershell"
+	"golang.org/x/exp/slices"
+)
+
+// Generic action function that may return an error
+type InvokeFn func() error
+
+// The type of command hooks. Supported values are 'pre' and 'post'
+type HookType string
+
+const (
+	// Executes pre-command hooks
+	HookTypePre HookType = "pre"
+	// Execute post-command hooks
+	HookTypePost HookType = "post"
+)
+
+// CommandHooks enable support to invoke integration scripts before & after commands
+// Scripts can be invoked at the project or service level or
+type CommandHooks struct {
+	commandRunner exec.CommandRunner
+	console       input.Console
+	cwd           string
+	scripts       map[string]*ScriptConfig
+	envVars       []string
+}
+
+// NewCommandHooks creates a new instance of CommandHooks
+func NewCommandHooks(
+	commandRunner exec.CommandRunner,
+	console input.Console,
+	scripts map[string]*ScriptConfig,
+	cwd string,
+	envVars []string,
+) *CommandHooks {
+	if cwd == "" {
+		osWd, err := os.Getwd()
+		if err != nil {
+			panic(err)
+		}
+
+		cwd = osWd
+	}
+
+	return &CommandHooks{
+		commandRunner: commandRunner,
+		console:       console,
+		cwd:           cwd,
+		scripts:       scripts,
+		envVars:       envVars,
+	}
+}
+
+// Invokes an action run runs any registered pre or post script hooks for the specified command.
+func (h *CommandHooks) Invoke(ctx context.Context, commands []string, actionFn InvokeFn) error {
+	err := h.RunScripts(ctx, HookTypePre, commands)
+	if err != nil {
+		return fmt.Errorf("failing running pre command hooks: %w", err)
+	}
+
+	err = actionFn()
+	if err != nil {
+		return err
+	}
+
+	err = h.RunScripts(ctx, HookTypePost, commands)
+	if err != nil {
+		return fmt.Errorf("failing running pre command hooks: %w", err)
+	}
+
+	return nil
+}
+
+// Invokes any registered script hooks for the specified hook type and command.
+func (h *CommandHooks) RunScripts(ctx context.Context, hookType HookType, commands []string) error {
+	scripts := h.getScriptsForHook(hookType, commands)
```

Comment:
FWIW, I can imagine a world where we might add an `azd` command for the hook author to validate that their scripts are properly registered and show errors if any happened.

For that purpose, I can see that the loading of script configuration to be separate than running the scripts (different classes), and maybe we'd gain something by separating it out in-code as such.

Reply:
Let me take a look.  Part of me is not looking to overly optimize until we need to support that use case.